### PR TITLE
move to a more standard DSL for F# HTML

### PR DESCRIPTION
**Pull request by dsyme**
This adjusts the HTML DSL for F# as described in https://github.com/dotnet/interactive/issues/687

Also includes #688 to trim back on default namespaces

Typical F# html code looks like this

```fsharp
open Html  // opens Microsoft.DotNet.Interactive.FSharp.FSharpKernelHelpers.Html

let register (df:DataFrame) (writer:TextWriter)
    let take = 20
    table [] [
      thead [] [
        th [] [i [ str "index"]]
        for c in df.Columns -> th [] [c.Name]
      ]
      tbody [] [
        for i in 0 .. min take (int df.Rows.Count - 1) do
          td [] [ object i ]
          for o in df.Rows.[int64 i] do
            td [] [ object o ]
      ]
    ]
    |> writer.Write
Formatter<DataFrame>.Register( (fun df writer -> register df writer), mimeType = "text/html")
```

TODO:
* [x] please check design
* [x] check the notebooks I adjusted by hand

### CODE REVIEW

**Comment by dsyme** [comment]
This is ready

File: src/Microsoft.DotNet.Interactive.FSharp.Tests/ApiTests.fs [hunk 1]
```diff
@@ -3,35 +3,39 @@
...
 
     [<Fact>]
-    member __.``inner HTML from content``() =
-        Assert.Equal("<div>d</div>", div.innerHTML("d").ToString())
+    member __.``inner HTML from string``() =
+        Assert.Equal("<div>d</div>", (div [] [str "d"]).ToString())
+
+    [<Fact>]
+    member __.``inner HTML from object``() =
+        Assert.Equal("<div>11</div>", (div [] [object 11]).ToString())
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.FSharp.Tests/ApiTests.fs:25]: A more interesting example here would be an object of a type that leads to additional formatting.

Comment (dsyme) [reply src/Microsoft.DotNet.Interactive.FSharp.Tests/ApiTests.fs:27]: I've added quite a number of more tests here

**Comment by dsyme** [comment]
I went through all the F# docs for formatting to make sure things were adding up.

Along the way I merged the "HTML.ipynb" into the "Displaying Outputs" and also renamed the notebook files so they don't have spaces in their names, to help me get the cross-references right.

I also added a fix and testing for #703.

**Comment by dsyme** [comment]
I also relented on a little on the question of which namespaces are open for F# and also included these by default:

    open System
    open System.IO
    open System.Text

I'll add this to the documentation.

**Comment by dsyme** [comment]
OK, I've added a document on the F# notebook programming model.

**Comment by dsyme** [comment]
@jonsequitur @colombod ready again

File: src/Microsoft.DotNet.Interactive.FSharp/FSharpHtml.fs [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+                | Text s -> encoder.Encode(writer, s)
+                | Tagged  p -> p.WriteTo(writer, encoder)
+                | Obj o -> encoder.Encode(writer, string o)
+
+    type HtmlAttribute = string * obj
+
+    [<AutoOpen>]
+    module HtmlElements =
+        type internal IMarker = interface end
+        let private f name (attrs: HtmlAttribute list) (content: IHtmlContent list) =
```
Comment (colombod) [inline src/Microsoft.DotNet.Interactive.FSharp/FSharpHtml.fs:34]: could we call ```f``` something like ```htmlElement``` or anything that could make it easier for contributors to identify?

File: src/Microsoft.DotNet.Interactive.FSharp/FSharpKernelExtensions.fs [hunk 3]
```diff
@@ -17,35 +17,53 @@ open XPlot.Plotly
...
-        let code = @"
-open System
-open System.Text
-open System.Threading.Tasks
-open System.Linq"
+        // F# has its own views on what namespaces are open by default in its scripting model
+        kernel
+
+    [<Extension>]
+    static member UseExtraNamespacesForTesting(kernel: FSharpKernelBase) =
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.FSharp/FSharpKernelExtensions.fs:48]: It looks like this should go in the test project.

File: src/Microsoft.DotNet.Interactive.FSharp/FSharpHtml.fs (hunk 2 above)
Comment (dsyme) [reply src/Microsoft.DotNet.Interactive.FSharp/FSharpHtml.fs:34]: Will do.  I will also remove the tags "html" and "body" which we never use in this context to reduce namespace pollution

File: src/Microsoft.DotNet.Interactive.FSharp/FSharpKernelExtensions.fs (hunk 3 above)
Comment (dsyme) [reply src/Microsoft.DotNet.Interactive.FSharp/FSharpKernelExtensions.fs]: Thanks, done

File: src/Microsoft.DotNet.Interactive.FSharp/FSharpHtml.fs (hunk 2 above)
Comment (dsyme) [reply src/Microsoft.DotNet.Interactive.FSharp/FSharpHtml.fs]: Thanks, done

**Comment by dsyme** [comment]
I've adjusted for code review (would love to get this one in as I'm needing it in documentation etc.)

**Pull request merged**